### PR TITLE
avoid python setup.py installdir error

### PR DESCRIPTION
**Pull request by naughtont3**
Fixes issue of ``make install`` failing with python bindings enabled.

### CODE REVIEW

**Comment by naughtont3** [comment]
Example error output...

```
bash:$ make install 
 ...<snip>...
make[4]: Entering directory '/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/
openpmix-git-br-master/bindings/python'
PYTHONPATH=$PYTHONPATH:/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpm
ix-git-br-master/_install/lib/python3.8/site-packages:/home/3t4/projects/pmix/sc
ratch/MODERN-WOLF/source/openpmix-git-br-master/_install/lib/python3.8/site-pack
ages \
	PMIX_BINDINGS_TOP_SRCDIR=/home/3t4/projects/pmix/scratch/MODERN-WOLF/sou
rce/openpmix-git-br-master PMIX_BINDINGS_TOP_BUILDDIR=/home/3t4/projects/pmix/sc
ratch/MODERN-WOLF/source/openpmix-git-br-master \
	/usr/bin/python3 ../../bindings/python/setup.py install --quiet --prefix
="/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/_ins
tall"
running install
error: can't create or remove files in install directory

The following error occurred while trying to add or remove files in the
installation directory:

    [Errno 2] No such file or directory: '/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/_install/lib/python3.8/site-packages/test-easy-install-3804489.writ
e-test'

The installation directory you specified (via --install-dir, --prefix, or
the distutils default setting) was:

    /home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/_install/lib/python3.8/site-packages/

This directory does not currently exist.  Please create it and try again, or
choose a different installation directory (using the -d or --install-dir
option).

make[4]: *** [Makefile:880: install-exec-local] Error 1
make[4]: Leaving directory '/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/bindings/python'
```

**Comment by rhc54** [comment]
Hmmm...something doesn't feel right here. I've been building the Python bindings repeatedly for months without problem, including a test run just now. The `make install` is supposed to make the target directory - and it indeed does so for me.

So the question is: why isn't it working for you? Can you paste your configure cmd line here?

**Comment by rhc54** [comment]
I just checked again with a totally new install location, and everything worked perfectly. I am using Python 3.10, but I wouldn't think that would matter. I'm using `pip 22.3.1` - what is your `pip` version?

**Comment by naughtont3** [comment]
this should be irrespective of the pip b/c it is just part of the `make install`.  But maybe you're getting at a difference in our setup.  I do not setup any special python env during the `make install` stage for ompi (or openpmix).  Maybe later when I run the python apps i do, but the error occurs during the install step.

**Comment by naughtont3** [comment]
Actually there is not a pip3 in the native system.  Python 3.8.10 is what is used for the install w/ Cython version 0.29.14.

**Comment by naughtont3** [comment]
I just tried again inside of a virtual environment and get the same result.  So not sure why you are not seeing this for your install?   Here's snip from my console log in case you see something different with your pathing.

```
make[3]: Entering directory '/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/bindings/python'
make[4]: Entering directory '/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/bindings/python'
PYTHONPATH=$PYTHONPATH:/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/_install/lib/python3.8/site-packages:/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/_install/lib/python3.8/site-packages \
	PMIX_BINDINGS_TOP_SRCDIR=/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master PMIX_BINDINGS_TOP_BUILDDIR=/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master \
	/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/ve/bin/python ../../bindings/python/setup.py install --quiet --prefix="/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/_install"
running install
error: can't create or remove files in install directory

The following error occurred while trying to add or remove files in the
installation directory:

    [Errno 2] No such file or directory: '/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/_install/lib/python3.8/site-packages/test-easy-install-94487.write-test'

The installation directory you specified (via --install-dir, --prefix, or
the distutils default setting) was:

    /home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/_install/lib/python3.8/site-packages/

This directory does not currently exist.  Please create it and try again, or
choose a different installation directory (using the -d or --install-dir
option).

make[4]: *** [Makefile:880: install-exec-local] Error 1
make[4]: Leaving directory '/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/bindings/python'
make[3]: *** [Makefile:728: install-am] Error 2
make[3]: Leaving directory '/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/bindings/python'
make[2]: *** [Makefile:722: install] Error 2
make[2]: Leaving directory '/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/bindings/python'
make[1]: *** [Makefile:694: install-recursive] Error 1
make[1]: Leaving directory '/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/bindings'
make: *** [Makefile:854: install-recursive] Error 1
(ve) node0:$ python3 --version
Python 3.8.10
(ve) node0:$ which python3
/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/ve/bin/python3
(ve) node0:$ which pip
/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/ve/bin/pip
(ve) node0:$ pip --version
pip 20.0.2 from /home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/ve/lib/python3.8/site-packages/pip (python 3.8)
(ve) node0:$ grep "pythondir =" bindings/python/Makefile
pkgpythondir = ${pythondir}/pmix
pythondir = ${prefix}/lib/python3.8/site-packages
(ve) node0:$ grep "prefix =" bindings/python/Makefile
exec_prefix = ${prefix}
prefix = /home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/_install
top_build_prefix = ../../
(ve) node0:$ ls /home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/_install
bin  etc  include  lib  share
(ve) node0:$ ls /home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/_install/lib
libpmix.la  libpmix.so  libpmix.so.2  libpmix.so.2.18.0  pmix
(ve) node0:$ ls /home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/_install/lib/python3.8
ls: cannot access '/home/3t4/projects/pmix/scratch/MODERN-WOLF/source/openpmix-git-br-master/_install/lib/python3.8': No such file or directory
(ve) node0:$
```

Could it is somehow be that you getting things put in the virtual environment's `site-packages` dir instead of the ${prefix} dir?  

I can see in above that when i look for `pip` it shows up under the ve's site-packages area.

**Comment by rhc54** [comment]
Yeah, nothing different using a CentOS VM:
```shell
Making install in bindings
make[1]: Entering directory `/opt/hpc/build/openpmix/bindings'
Making install in python
make[2]: Entering directory `/opt/hpc/build/openpmix/bindings/python'
make  install-am
make[3]: Entering directory `/opt/hpc/build/openpmix/bindings/python'
make[4]: Entering directory `/opt/hpc/build/openpmix/bindings/python'
PYTHONPATH=$PYTHONPATH:/opt/hpc/external/pmix/lib/python3.6/site-packages:/opt/hpc/external/pmix/lib64/python3.6/site-packages \
	PMIX_BINDINGS_TOP_SRCDIR=/opt/hpc/build/openpmix PMIX_BINDINGS_TOP_BUILDDIR=/opt/hpc/build/openpmix \
	/usr/bin/python3 ../../bindings/python/setup.py install --quiet --prefix="/opt/hpc/external/pmix"
running install
running bdist_egg
running egg_info
creating pypmix.egg-info
writing pypmix.egg-info/PKG-INFO
writing dependency_links to pypmix.egg-info/dependency_links.txt
writing requirements to pypmix.egg-info/requires.txt
writing top-level names to pypmix.egg-info/top_level.txt
writing manifest file 'pypmix.egg-info/SOURCES.txt'
reading manifest file 'pypmix.egg-info/SOURCES.txt'
writing manifest file 'pypmix.egg-info/SOURCES.txt'
installing library code to build/bdist.linux-aarch64/egg
running install_lib
running build_ext
creating build/bdist.linux-aarch64
creating build/bdist.linux-aarch64/egg
copying build/lib.linux-aarch64-3.6/pmix.cpython-36m-aarch64-linux-gnu.so -> build/bdist.linux-aarch64/egg
creating stub loader for pmix.cpython-36m-aarch64-linux-gnu.so
byte-compiling build/bdist.linux-aarch64/egg/pmix.py to pmix.cpython-36.pyc
creating build/bdist.linux-aarch64/egg/EGG-INFO
copying pypmix.egg-info/PKG-INFO -> build/bdist.linux-aarch64/egg/EGG-INFO
copying pypmix.egg-info/SOURCES.txt -> build/bdist.linux-aarch64/egg/EGG-INFO
copying pypmix.egg-info/dependency_links.txt -> build/bdist.linux-aarch64/egg/EGG-INFO
copying pypmix.egg-info/not-zip-safe -> build/bdist.linux-aarch64/egg/EGG-INFO
copying pypmix.egg-info/requires.txt -> build/bdist.linux-aarch64/egg/EGG-INFO
copying pypmix.egg-info/top_level.txt -> build/bdist.linux-aarch64/egg/EGG-INFO
writing build/bdist.linux-aarch64/egg/EGG-INFO/native_libs.txt
creating dist
creating 'dist/pypmix-6.0.0-py3.6-linux-aarch64.egg' and adding 'build/bdist.linux-aarch64/egg' to it
removing 'build/bdist.linux-aarch64/egg' (and everything under it)
Creating /opt/hpc/external/pmix/lib64/python3.6/site-packages/site.py
Processing pypmix-6.0.0-py3.6-linux-aarch64.egg
creating /opt/hpc/external/pmix/lib64/python3.6/site-packages/pypmix-6.0.0-py3.6-linux-aarch64.egg
Extracting pypmix-6.0.0-py3.6-linux-aarch64.egg to /opt/hpc/external/pmix/lib64/python3.6/site-packages
Adding pypmix 6.0.0 to easy-install.pth file

Installed /opt/hpc/external/pmix/lib64/python3.6/site-packages/pypmix-6.0.0-py3.6-linux-aarch64.egg
Processing dependencies for pypmix==6.0.0
Searching for Cython==0.28.5
Best match: Cython 0.28.5
Adding Cython 0.28.5 to easy-install.pth file
Installing cygdb script to /opt/hpc/external/pmix/bin
Installing cython script to /opt/hpc/external/pmix/bin
Installing cythonize script to /opt/hpc/external/pmix/bin

Using /usr/lib64/python3.6/site-packages
Finished processing dependencies for pypmix==6.0.0
make[4]: Nothing to be done for `install-data-am'.
make[4]: Leaving directory `/opt/hpc/build/openpmix/bindings/python'
make[3]: Leaving directory `/opt/hpc/build/openpmix/bindings/python'
make[2]: Leaving directory `/opt/hpc/build/openpmix/bindings/python'
```
What I find disturbing is that your output seems to be missing a lot of stuff...and of course, you aren't getting the directory tree constructed.

This was done with Python 3.6, python3 and pip3.

**Comment by rhc54** [comment]
FWIW: I'm not operating in a virtualenv, so pip and python are both the site versions. I typically do operate in a VE at home, but not on my notebook and not in the VM.

**Comment by naughtont3** [comment]
OK, did a bit more testing and "learned" a couple thing.

    Do not see make install problem on Centos 7.9 VM (but did need pip3 install --user Cython to install a python3 module)
    Do not see make install problem if you remove the installdir from PYTHONPATH when running the python3 setup.py... during make install. (thanks Wael for spotting this diff)
    But consistently saw make install on Ubuntu (and SLES olcf systems) unless add the mkdir -p bit

But it looks like you merged this. But figured I'd add a bit more info to ticket for trail

**Comment by naughtont3** [comment]
So I did a few more tests:
 1. Problem does not occur on Centos 7.9
 2. Problem does occur on Ubuntu and SLES
 3. Seems there is no harm on Centos to add the `$(MKDIR_P)`
 4. Fixes `make install` failure on Ubuntu and SLES by adding `$(MKDIR_P)`

Regardless, you need to update your PYTHONPATH afterward to this install directory to uses this install.  Also, the install is always under the `${prefix}` area so if not already there, you will want it there when the `make install` completes with the python setup.py.

Bottom line, I don't see any harm in adding the `$(MKDIR_P)` and things fail on some platforms w/o it.

That's my pitch. :-)
